dm/pkg/shardddl/optimism/lock.go

```diff
@@ -75,30 +82,94 @@ type Lock struct {
-		DownSchema: downSchema,
-		DownTable:  downTable,
-		joined:     joined,
-		tables:     make(map[string]map[string]map[string]schemacmp.Table),
-		done:       make(map[string]map[string]map[string]bool),
-		synced:     true,
-		versions:   make(map[string]map[string]map[string]int64),
-		columns:    make(map[string]map[string]map[string]map[string]DropColumnStage),
+		cli:            cli,
+		ID:             id,
+		Task:           task,
+		DownSchema:     downSchema,
+		DownTable:      downTable,
+		joined:         joined,
+		tables:         make(map[string]map[string]map[string]schemacmp.Table),
+		done:           make(map[string]map[string]map[string]bool),
+		synced:         true,
+		versions:       make(map[string]map[string]map[string]int64),
+		columns:        make(map[string]map[string]map[string]map[string]DropColumnStage),
+		downstreamMeta: downstreamMeta,
 	}
 	l.addTables(tts)
 	metrics.ReportDDLPending(task, metrics.DDLPendingNone, metrics.DDLPendingSynced)
-
+	// pre join because tables may have different schema at the beginning
+	l.joinTable()
 	return l
 }
 
+// FetchTableInfos fetch all table infos for a lock.
+func (l *Lock) FetchTableInfos(task, source, schema, table string) (*model.TableInfo, error) {
+	if l.downstreamMeta == nil {
+		return nil, terror.ErrMasterOptimisticDownstreamMetaNotFound.Generate(task)
+	}
+
+	db, err := conn.DefaultDBProvider.Apply(l.downstreamMeta.dbConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), dbutil.DefaultTimeout)
+	defer cancel()
+
+	query := `SELECT table_info FROM ` + dbutil.TableName(l.downstreamMeta.meta, cputil.SyncerCheckpoint(task)) + ` WHERE id = ? AND cp_schema = ? AND cp_table = ?`
+	row := db.DB.QueryRowContext(ctx, query, source, schema, table)
+	if row.Err() != nil {
+		return nil, terror.ErrDBExecuteFailed.Delegate(row.Err(), query)
+	}
+	var tiBytes []byte
+	if err := row.Scan(&tiBytes); err != nil {
+		return nil, terror.ErrDBExecuteFailed.Delegate(err, query)
+	}
+	var ti *model.TableInfo
+	if bytes.Equal(tiBytes, []byte("null")) {
+		log.L().Warn("null table info", zap.String("query", query), zap.String("source", source), zap.String("schema", schema), zap.String("table", table))
+		return nil, terror.ErrMasterOptimisticDownstreamMetaNotFound.Generate(task)
+	}
+	if err := json.Unmarshal(tiBytes, &ti); err != nil {
+		return nil, err
+	}
+	return ti, nil
+}
+
+// joinTable join tables for a lock and update l.joined.
+func (l *Lock) joinTable() {
+	var (
+		joined     = l.joined
+		firstTable = true
+	)
+	for _, schemaTables := range l.tables {
+		for _, tables := range schemaTables {
+			for _, ti := range tables {
+				if firstTable {
+					joined = ti
+					firstTable = false
+				} else {
+					newJoined, err := joined.Join(ti)
+					if err != nil {
+						log.L().Error(fmt.Sprintf("fail to join table info %s with %s", joined, ti), zap.String("lockID", l.ID), log.ShortError(err))
```

Comment:
just a question why can we mute this join err here ?

Reply:
`joinTable` only use when create a lock. User may create a task with conflict upstream. e.g.
```
tb1(a int, b int);
tb2(a int, b varchar(4));
```
In this situation, init joined will error, but we can just ignore it. #4297 have a further implove.